### PR TITLE
[Dev] Allow FP8/FP4 with AbsorbedMLA up-projection recompute

### PR DESCRIPTION
**Pull request by niyunsheng**
## Summary

`AbsorbedMLASelfAttention` asserted that quantization was unsupported whenever `"mla_up_proj"` selective recompute was enabled, blocking FP8 and FP4 training for any model that needs that recompute to fit in memory. The assertion arrived with the initial absorbed-MLA implementation (#3193 / #3198) as a not-yet-validated placeholder rather than a guard for a known failure.

The quantized replay is safe for the same reason it is in `MLASelfAttention`, which runs the identical `CheckpointWithoutOutput` pattern with no such assertion (`multi_latent_attention.py`):

- The checkpoint records the forward recipe and amax state (`ctx.fp8` /
  `ctx.fp8_recipe` under `activation_recompute_forward(recompute_phase=False)`) and replays under the recorded `fp8_autocast` with `recompute_phase=True`.
- The only quantized operation inside `qkv_up_proj_and_rope_apply` is the Q up projection — the same module set MLA already replays.
- The absorption einsum reads the K up-projection weight directly rather than through the checkpoint context. That weight is a persistent parameter: under
  `fp8_param_gather` it is remapped to TE quantized storage that outlives the backward pass (`param_and_grad_buffer.py`), and the only path that frees param storage is explicit weight CPU offload. So the forward and the replay observe the same values.

`DeepSeekV4HybridAttention` already runs the same pattern without the assertion.

## Test plan

Added `test_fp8_up_proj_recompute_parity`, which drives two identically-initialized modules — one recomputing the up projection, one not — over the same inputs under `get_fp8_context`, and requires **bitwise-identical** (`atol=0, rtol=0`) outputs and per-parameter gradients. Recompute is meant to be an exact replay, so any drift would indicate a different quantization scale or a stale weight; a tolerance-based comparison would hide exactly the failures this guards against.
A bf16 reference run guards the parity check against passing vacuously: the same weights go through an `fp8=False` module, and the FP8 output must track it closely and differ from it bitwise — which is only possible if quantization actually engaged.

Coverage: `tensorwise` / `mxfp8` / `blockwise` recipes x `sbhd` / `thd` layouts x combined / split K-V up-projection module specs.

Run on GB200 (mxfp8 needs Blackwell, so nothing is capability-skipped there):

```
pytest tests/unit_tests/transformer/experimental_attention_variant/test_absorbed_mla.py
====================== 76 passed, 558 warnings in 52.58s =======================
```

76 = 64 pre-existing `test_functionality` cases (unchanged, no BF16 regression) + 12 new FP8 cases, 0 skipped.

### CODE REVIEW

File: tests/unit_tests/transformer/experimental_attention_variant/test_absorbed_mla.py [hunk 1]
```diff
@@ -432,3 +439,123 @@ def _calculate_tensor_similarity(x, y):
...
 
     Utils.destroy_model_parallel()
+
+
+# Hopper = SM 9.0, Blackwell = SM 10.0+. mxfp8 needs Blackwell.
+_IS_BLACKWELL = torch.cuda.is_available() and torch.cuda.get_device_capability()[0] >= 10
+
+_QUANT_RECIPES = [
+    pytest.param("tensorwise", id="fp8-tensorwise"),
+    pytest.param("delayed", id="fp8-delayed"),
```
Comment (kunlunl) [inline tests/unit_tests/transformer/experimental_attention_variant/test_absorbed_mla.py]: Better to test blockwise scaling rather than delayed. Delayed scaling is almost not used in anywhere.

File: tests/unit_tests/transformer/experimental_attention_variant/test_absorbed_mla.py [hunk 2]
```diff
@@ -432,3 +439,123 @@ def _calculate_tensor_similarity(x, y):
...
+    if state_dict is not None:
+        model.load_state_dict(state_dict)
+    return model
+
+
+@pytest.mark.skipif(not fp8_available, reason=reason_for_no_fp8)
+@pytest.mark.parametrize("fp8_recipe", _QUANT_RECIPES)
+@pytest.mark.parametrize("qkv_format", ['sbhd', 'thd'])
+@pytest.mark.parametrize("combined_kv_up_projection", [True, False])
+def test_fp8_up_proj_recompute_parity(
```
Comment (kunlunl) [inline tests/unit_tests/transformer/experimental_attention_variant/test_absorbed_mla.py]: This test only tests fp8 recompute against fp8 no-recompute. Better to test fp8 recompute vs bf16 baseline as other tests.

**Comment by niyunsheng** [comment]
> **Non-blocking suggestion (test coverage):** The code path also enables FP4 (`self.config.fp8 or self.config.fp4`) and the PR title says "FP8/FP4," but the test only exercises FP8 recipes; the test's guarantee is effectively FP8-scoped. FP4 shares the exact same code path and requires TE ≥2.7 + capable hardware, so deferring it is reasonable — worth a note in the test docstring or a follow-up.

Good catch — I will add `nvfp4` coverage since I have Blackwell available.

**Review by kunlunl — APPROVED** [review]

**Review by FDecaYed — APPROVED** [review]
LGTM

**Comment by kunlunl** [comment]
/ok to test [ece6447](https://github.com/NVIDIA/MegatronLM/pull/6178/commits/ece6447f91f7548cd9d313eee97eed3d7c2f0c58)

**Comment by niyunsheng** [comment]
The failing bucket `tests/unit_tests/transformer/**/*.py` - latest is unrelated to this PR: the only failure is [test_te_layers_batch_invariant.py::test_te_column_parallel_linear_batch_invariant_randomized](https://github.com/NVIDIA/Megatron-LM/actions/runs/30885006091/job/91934222450?pr=6178#step:4:11874) with RuntimeError. But this PR only touches `absorbed_mla.py` and its test file.

Could someone re-run the failed job (or re-trigger with `/ok to test`)? Thanks! @buptzyb @kunlunl @FDecaYed

**Comment by buptzyb** [comment]
It failed again. Are you sure it's not related to your changes?

**Comment by niyunsheng** [comment]
> It failed again. Are you sure it's not related to your changes?

@buptzyb CI is green now. Root cause was an import side effect: the test imported helpers from `tests.unit_tests.determinism.utils`, which transitively imports `torch.testing._internal.common_utils` and freezes torch backend global flags (`disable_global_flags()`), deterministically breaking `test_te_layers_batch_invariant.py` later in the same session. Fixed in `6648492` by inlining the two RNG helpers so `torch.testing._internal` is never loaded in this bucket.
